community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java

```diff
@@ -626,7 +623,17 @@ private void afterRollback()
     private void release()
     {
         locks.releaseAll();
-        pool.release( this );
+        if ( terminated )
+        {
+            // This transaction has been externally marked for termination.
+            // Just dispose of this transaction and don't return it to the pool.
+            dispose();
+        }
+        else
+        {
+            // Return this instance to the pool so that another transaction may use it.
+            pool.release( this );
+        }
```

Comment:
Terminating transactions was originally intended for cancelling long-running queries. This adds another meaning to termination, and a behaviour that isn't always needed. Perhaps not a big problem in practice, but we could alternatively change the boolean to an enum that is either CANCEL or DISPOSE, so we keep the single-volatile-write signalling and exact semantics for the behaviour we want.